src/java/org/apache/cassandra/dht/RangeStreamer.java

```diff
@@ -373,4 +369,34 @@ public StreamResultFuture fetchAsync()
 
         return streamPlan.execute();
     }
+
+    public boolean areAllRangesPresent()
+    {
+        boolean allPresent = true;
+        for (Map.Entry<String, Map.Entry<InetAddress, Collection<Range<Token>>>> entry : toFetch.entries())
+        {
+            Collection<Range<Token>> ranges = removeAvailableRanges(entry, false);
+            allPresent &= ranges.isEmpty();
+        }
+        return allPresent;
+    }
+
+    @VisibleForTesting
+    Set<Range<Token>> removeAvailableRanges(Map.Entry<String, Map.Entry<InetAddress, Collection<Range<Token>>>> entry, boolean log)
+    {
+        String keyspace = entry.getKey();
+        Collection<Range<Token>> ranges = ImmutableSet.copyOf(entry.getValue().getValue());
+
+        // filter out already streamed ranges
+        Set<Range<Token>> availableRanges = stateStore.getAvailableRanges(keyspace, StorageService.getPartitioner());
+        Set<Range<Token>> unavailableRanges = ranges.stream()
+                                                    .filter(range -> !availableRanges.contains(range))
+                                                    .collect(Collectors.toSet());
+        entry.getValue().getValue().removeAll(availableRanges);
```

Comment:
Mostly a nit, but I think it's slightly more idiomatic to just do:

HashSet<> modifiableSet = new HashSet<>(entry.getValue().getValue());
modifiableSet.removeAll(availableRanges);
return ImmutableSet.copyOf(modifiableSet);

Reply:
or with streams:

entry.getValue().getValue().stream()
.filter(x -> !availableRanges.contains(x))
.toSet();